src/compiler/checker.ts

```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         for (let i = 0; i < paramCount; i++) {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {
+                targetType = mapType(targetType, t => {
+                    if (!isTupleType(t) || isTypeIdenticalTo(sourceType, t)) {
+                        return t;
+                    }
+
+                    const elementTypes: Type[] = [];
+                    const elementFlags: ElementFlags[] = [];
+
+                    for (let i = 0; i < getTypeReferenceArity(sourceType); i++) {
+                        elementTypes.push(getTupleElementType(t, i)!);
+                        elementFlags.push(sourceType.target.elementFlags[i]);

```

Comment:
That's an excellent question. At first, I wasn't sure if this is correct - it has been a while since I worked on this. I rechecked that and this is intentional. The implied source tuple type, which represents parameters of the source signature, might only have fixed elements - while the target tuple type might be declared using rest. For example:
```ts
const f1: (...args: [number, string, ...boolean[]] | [string, number, ...boolean[]]) => void = (a, b, c) => {};
```

Here, we want to compare those *contravariant* tuples:
```ts
type Target = [number, string, boolean] | [string, number, boolean]
type Source = [a: string | number, b: string | number, c: boolean]
```

However, at the moment - I'm not entirely sure if all "transitions" that can happen here are OK. I need to analyze this again and think if it's always OK to, for example, create an optional element from some required elements.

Reply:
I'm trying to review this again, and I'm trying to think if using the source tuple type's element flags to construct the new target tuple types is ok. I was trying to think through the case where the source element is optional, so the new target element will be too. I think this makes sense and is OK to do when we're comparing function signatures strictly, i.e. when we'll compare parameter types contravariantly. However, if we're comparing *method* signatures, we'll end up comparing the parameter types also covariantly, and in that case I don't know if it makes sense for the new target type element to be marked as optional for that comparison.

A perhaps convoluted example I came up with to try and show this:
```ts
// Counter-example to making target element optional when source element is optional?
type NowSource<T extends boolean[]> = [...rest: [boolean, ...T, n?: number]];
type NowTarget<T extends boolean[]> = [...rest: [boolean, ...T, number | string]];

class A {
  method<T extends boolean[]>(...args: NowTarget<T>): void {}
}

class B extends A {
  method<T extends boolean[]>(...args: NowSource<T>): void {} // Errors on main, no error on PR
}

declare let func: <T extends boolean[]>(...args: NowTarget<T>) => void;
declare let gunc: <T extends boolean[]>(...args: NowSource<T>) => void;

func = gunc; // Errors on main and PR, ok
```